src/asm3/animal.py

```diff
@@ -3847,8 +3851,169 @@ def update_diary_linkinfo(dbo: Database, animalid: int, a: ResultRow = None, dia
+        ", ".join(releventmovementtypes),
+        ") ORDER BY MovementDate desc, ReturnDate desc"
+    ])
+    movementrows = dbo.query(query, (animalid,))
+    animallocations = dbo.query("SELECT ID, Date, MovementID, IsDeath, ToLocationID, ToUnit, PrevAnimalLocationID FROM animallocation WHERE AnimalID = ? ORDER BY DATE desc", (animalid,))
+    ## Detect and remove impossible internal movements
+    indates = [entrydate]
+    outdates = []
+    if movementrows:
+        for row in movementrows:
+            outdates.append(asm3.i18n.remove_time(row["MOVEMENTDATE"]))
+            if row["RETURNDATE"]:
+                indates.append(asm3.i18n.remove_time(row["RETURNDATE"]))
+    if deceaseddate:
+        outdates.append(deceaseddate)
+    validrows = []
+    indate = indates[0]
+    indatecount = 0
+    outdatecount = 0
+    while indatecount < len(indates):
+        outdate = None
+        for outdate in outdates[outdatecount:]:
+            if outdate >= indate:
+                outdatecount = outdatecount + 1
+                break
+        indatecount = indatecount + 1
+        for row in animallocations:
+            if outdate:
+                if row["MOVEMENTID"] == 0 and row["ISDEATH"] == 0 and asm3.i18n.remove_time(row["DATE"]) >= indate and asm3.i18n.remove_time(row["DATE"]) <= outdate:
+                    if row not in validrows:
+                        validrows.append(row)
+            elif row["MOVEMENTID"] == 0 and row["ISDEATH"] == 0 and asm3.i18n.remove_time(row["DATE"]) >= indate:
+                    if row not in validrows:
+                        validrows.append(row)
+    for row in animallocations.copy():
+        if row not in validrows:
+            dbo.delete("animallocation", row["ID"], username)
+            animallocations.remove(row)
+            rowid = row["ID"]
+            rowprevlocationid = row["PREVANIMALLOCATIONID"]
+            dbo.execute(
+                "UPDATE animallocation SET PrevAnimalLocationID = ? WHERE AnimalID = ? AND PrevAnimalLocationID = ?",
+                (rowprevlocationid, animalid, rowid)
+            )
+    
+    ## Detect and delete obsolete external movements
+    obsoletemovements = []
+    for row in animallocations:
+        if row["MOVEMENTID"]:
+            movementfound = False
+            for movement in movementrows:
+                if row["MOVEMENTID"] == movement["ID"]:
+                    movementfound = True
+                    break
+            if not movementfound:
+                obsoletemovements.append(row)
+    for row in obsoletemovements:
+        dbo.delete("animallocation", row["ID"], username)
+        animallocations.remove(row)
+        rowid = row["ID"]
+        rowprevlocationid = row["PREVANIMALLOCATIONID"]
+        dbo.execute(
+            "UPDATE animallocation SET PrevAnimalLocationID = ? WHERE AnimalID = ? AND PrevAnimalLocationID = ?",
+            (rowprevlocationid, animalid, rowid)
+        )
+
+    ## Sync external movements
```

Comment:
More info on what this is doing and how it's doing it? Looks like the first part is checking and updating the date from movement to exit animal location, and adding a new one if it can't find one